src/search-and-replace-dialog.jsx

```diff
@@ -0,0 +1,273 @@
+                replaced = true;
+                globalIndex ++;
+                return replaceString;
+            } else {
+                globalIndex ++;
+                return match;
+            }
+        });
+
+        if (!replaced) {
+            answerArea.options.content =
+                answerArea.options.content.replace(regex, match => {
+                    if (!replaced && globalIndex === searchIndex) {
+                        replaced = true;
+                        globalIndex ++;
+                        return replaceString;
+                    } else {
+                        globalIndex ++;
+                        return match;
+                    }
+                });
+        }
+
+        if (!replaced) {
+            hints.forEach(hint => {
+                if (replaced) {
+                    return;
+                }
+                hint.content = hint.content.replace(regex, match => {
+                    if (!replaced && globalIndex === searchIndex) {
+                        replaced = true;
+                        globalIndex ++;
+                        return replaceString;
+                    } else {
+                        globalIndex ++;
+                        return match;
+                    }
+                });
+            });
+        }
+
+        var searchResultCount = this.state.searchResultCount;
+
+        if (replaceString.indexOf(searchString) === -1) {
+            // normal case: replaceString doesn't contain searchString so we
+            // can decrement searchResultCount because we just replaced one 
+            // occurence of searchString
+            searchResultCount --;
+        } else {
+            // edge case: replaceString contains one or more instances of
+            // searchString which means the searchResultCount will either 
+            // stay the same if the number of instances is one or increase
+            // if it's more than one
+            var matches = replaceString.match(regex);
+            if (matches) {
+                searchResultCount += matches.length - 1;
+            }
+        }
+
+        if (searchIndex >= searchResultCount) {
+            searchIndex = searchResultCount - 1;
+        }
+
+        this.setState({ searchIndex, searchResultCount });
+        this.props.onSearchChange({ searchIndex });
+        this.props.onDocumentChange({ question, answerArea, hints });
+    },
+
+    style: {
+        padding: 10,
+        position: 'fixed',
+        right: 0,
+        top: 0,
+        width: 300,
+        backgroundColor: '#EEE',
+        border: 'solid 1px #DDD',
+        zIndex: 100
+    },
+
+    labelStyle: {
+        display: 'inline-block',
+        float: 'left',
+        clear: 'left',
+        textAlign: 'right',
+        width: 60,
+        marginRight: 8
+    },
+
+    inputStyle: {
+        display: 'inline-block',
+        float: 'left',
+        width: 220,
+        boxSizing: 'border-box'
+    },
```

Comment:
I'm not super happy about all the magic numbers in here.  I'm going to rewrite this using flexbox so that it's more flexible.